waiter/src/waiter/scheduler/kubernetes.clj

```diff
@@ -0,0 +1,483 @@
+         service-id (get-in pod [:metadata :annotations :waiter/service-id])]
+     (str service-id \. instance-suffix \- restart-count))))
+
+(defn- track-failed-instances [live-instance pod]
+  (when-let [newest-failure (get-in pod [:status :containerStatuses 0 :lastState :terminated])]
+    (let [service-id (:service-id live-instance)
+          instance-id (:id live-instance)
+          instance-x (pod->instance-id pod \X)
+          newest-failure-start-time (-> newest-failure :startedAt timestamp-str->datetime)
+          failures (-> instance-x->failed-instances-transient-store
+                       deref (get instance-x))
+          prev-failure-start-time (some-> failures last :started-at)]
+      ;; FIXME - this is racey! we can end up reporting the same failed instances twice.
+      (when (or (nil? prev-failure-start-time)
+                (neg? (compare prev-failure-start-time newest-failure-start-time)))
+        (let [failure-flags (cond
+                              (= "OOMKilled" (:reason newest-failure)) #{:memory-limit-exceeded}
+                              :else #{})
+              ;; TODO - Look at events for messages about liveness probe failures:
+              ;; /api/v1/namespaces/<ns>/events?fieldSelector=involvedObject.namespace=<ns>,involvedObject.name=<instance-id>,reason=Unhealthy
+              ;; (-> event :message (string/starts-with? "Liveness probe failed:")) #{:never-passed-health-checks}
+              ;; For now, we just assume any SIGKILL (137) with the default "Error" reason was a livenessProbe kill.
+              killed-by-k8s? (and (= 137 (:exitCode newest-failure))
+                                  (= "Error" (:reason newest-failure)))
+              newest-failure-instance (merge live-instance
+                                             ; to match the behavior of the marathon scheduler,
+                                             ; don't include the exit code in failed instances that were killed by k8s
+                                             (when-not killed-by-k8s?
+                                               {:exit-code (:exitCode newest-failure)})
+                                             {:flags failure-flags
+                                              :healthy? false
+                                              :id (pod->instance-id pod (count failures))
+                                              :started-at newest-failure-start-time})]
+          (doseq [[id store] [[service-id service-id->failed-instances-transient-store]
+                              [instance-x instance-x->failed-instances-transient-store]]]
+            (swap! store update-in [id] (comp vec conj) newest-failure-instance)))))))
+
+(defn- pod->ServiceInstance
+  [pod]
+  ;; FIXME - catch exceptions and return nil
+  (let [port0 (get-in pod [:spec :containers 0 :ports 0 :containerPort])]
+  (scheduler/make-ServiceInstance
+    {:app-name (get-in pod [:metadata :labels :app])
```

Comment:
Nitpick: rename `app-name` to `service-name`

Reply:
`app-name` → `k8s-name`